### PR TITLE
Open Domain Checkout web-view in Site Creation

### LINKED ISSUE #17905: Add checkout functionality and UI
The flow for paid domains should continue with a new step: Checkout.
This task is to implement and show this step in a webview.
See internal ref: `pc8HXX-ZF-p2` for API integration details.

#### Subtasks
- [x] #18086
  - [x] #18087
- [x] #18108
  - [x] wordpress-mobile/WordPress-FluxC-Android#2676
- [x] #18177
  - [x] ~~Update domain checkout web-view to be reusable in site creation~~ → This wasn't needed
  - [x] Create shopping cart for domain purchase
- [x] #18255

New functionality:
Integrate the domain checkout step in the site creation flow after the user taps `Purchase domain`:

1. Call API to create site
2. Call API to create shopping cart
3. Open `wordpress.com/checkout/{new_site_subdomain}.wordpress.com` in a webview
4. After the purchase, the final preview screen of the site creation flow should be displayed
   - The implementation should make sure the preview screen doesn't call the API to create a new site, as that should have already happened.

### Design

---
> **Note**:
> - Other changes are out of scope and part of other tasks
> - All development should be behind [the feature flag](https://github.com/wordpress-mobile/WordPress-Android/issues/17896)

### PR DESCRIPTION
**Pull request by ovitrif**
Resolves partially #17905

This PR:

- ★ Implements the logic to reuse the existing domain registration checkout and open it in a web-view
- ⊕ Adds API call to create shopping cart for the paid domain product before checkout
- ⊜ Refactors the vm for the progress screen to support returning to it after back navigation from checkout

#### Prerequisite

<details><summary>◐ Toggle the <code>SiteCreationDomainPurchasingFeatureConfig</code> flag</summary>

1. Go to `Me` → `App Settings` → `Debug settings`
2. Scroll to the `Features in development` section
3. Tap on the item corresponding to the flag
4. Tap `RESTART THE APP` button
</details>

### ● Treatment Variation

- **Verify** Checkout is opened and filled in with correctly for domain purchases in site creation
  - **Expect** discounts to be applied on Checkout screen
- **Verify** Back navigation from Checkout is supported
  - **Expect** to be able to create a site when resuming with a free domain
  - **Expect** the previously created wpcom blog to be reused when resuming with same domain
  - **Expect** Checkout is opened and filled in correctly when resuming with a different paid domain

### ○ Control Variation

- **Verify** no regression in the site creation flow: expect no issues when creating a new site with a free domain

### Preview

| ● |
| - |
| <video src="https://user-images.githubusercontent.com/4588074/228221635-f6b76ecc-67e3-4827-9203-0b32227c569a.mp4" style="max-width: 315px;"></video> |

> **Note**
> The recording shows a preview of all tests passing for the **● Treatment Variation**.

## Regression Notes
1. Potential unintended areas of impact
  `SITE CREATION` → `PROGRESS`

2. What I did to test those areas of impact (or what existing automated tests I relied on)
  Manual testing and unit testing

3. What automated tests I added (or what prevented me from doing so)
  Unit tests in `SiteCreationMainVMTest` and `SiteCreationProgressViewModelTest`

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/SiteCreationActivity.kt
Comment (thomashorta) [inline WordPress/src/main/java/org/wordpress/android/ui/sitecreation/SiteCreationActivity.kt]: ❓ Any particular reason for not using something like the snippet below?
```kotlin
is CreatedButNotFetched, is DomainRegistrationPurchased -> {
    // code here
}
```

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/progress/SiteCreationProgressViewModel.kt
Comment (thomashorta) [inline WordPress/src/main/java/org/wordpress/android/ui/sitecreation/progress/SiteCreationProgressViewModel.kt:95]: ❓ Just a question/confirmation here: both variables have the same name (`siteCreationState`), is this line using the correct one? (it's using from the parameter)

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/progress/SiteCreationProgressViewModel.kt
Comment (thomashorta) [inline WordPress/src/main/java/org/wordpress/android/ui/sitecreation/progress/SiteCreationProgressViewModel.kt]: 🔍 Nitpicking: I feel like this event should be tracked closer to the action that actually starts the WebView like [here](https://github.com/wordpress-mobile/WordPress-Android/pull/18177/files#diff-56921dabf3cd181875890b198607814bd8cd3c944dd8e3d46cab28d6e63b738fR300) (`showDomainCheckout` from the other VM), since it seems a bit out of place to track "webViewViewed" in the `onCartCreated` event. Wdyt?

**Review by thomashorta — COMMENTED** [review]
Overall looks good to me, I just added a few points before I approve.

Also, everything worked while testing but the "sale" domain price shown in the domain picker didn't match the checkout page, see screenshots:
Picker | Checkout
--- | ---
domain-price-sale-picker | domain-price-sale-checkout

Also, another feedback unrelated to this PR's work is that the UI is a bit confusing since the selected domain is not highlighted in any way.

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/SiteCreationActivity.kt
Comment (ovitrif) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/SiteCreationActivity.kt]: Good question @thomashorta, I might have missed to remodel this after I confirmed that everything works via testing.
Resolved in: 61198775

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/progress/SiteCreationProgressViewModel.kt
Comment (ovitrif) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/progress/SiteCreationProgressViewModel.kt:95]: I'm aware of it and that's why I used `this.` when I wanted the value cached in the VM 👍🏻

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/progress/SiteCreationProgressViewModel.kt
Comment (ovitrif) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/progress/SiteCreationProgressViewModel.kt]: Good point, I reckon I forgot it since the initial copy-pasta "work" 😂.
Resolved in d7a4cd31

**Review by ovitrif — COMMENTED** [review]
@thomashorta Thanks a lot for the review and your remarks 🙇🏻 

I added changes to resolve all concerns you had over the code.

**Comment by ovitrif** [comment]
> Also, another feedback unrelated to this PR's work is that the UI is a bit confusing since the selected domain is not highlighted in any way.

👋🏻 @thomashorta 
I totally agree with you on this one, unfortunately we still lack a UI design for highlighting the selected domain, and we are already at the 4th design iteration… 🤦🏻 

Today I collaborated with the developers of the project team and came up with a UI design proposal ourselves.

However, the implementation was already foreseen to be tackled separately, in task:
- #17901

**Comment by ovitrif** [comment]
> Also, everything worked while testing but the "sale" domain price shown in the domain picker didn't match the checkout page, see screenshots:

Nice find 🙇🏻 @thomashorta .

Thank you for testing discounts with a different locale. I never encountered this issue with €, but also never tested before with other currencies. I plan to fix this before merging the PR.

**Review by thomashorta — APPROVED** [review]
> Thank you for testing discounts with a different locale. I never encountered this issue with €, but also never tested before with other currencies. I plan to fix this before merging the PR.

As we talked, I think this shouldn't block this PR and has already been added to the tasks in the main issue, so I am approving this one!

Thanks for the contribution and for addressing the review! :shipit:

**Comment by ovitrif** [comment]
> > Thank you for testing discounts with a different locale. I never encountered this issue with €, but also never tested before with other currencies. I plan to fix this before merging the PR.
> 
> As we talked, I think this shouldn't block this PR and has already been added to the tasks in the main issue, so I am approving this one!
> 
> Thanks for the contribution and for addressing the review! :shipit:


Posting this for reference & backtracking reasons:
After our DM chat, we concluded it's ok to solve this as part of the main task, where I added a subtask for it:
- https://github.com/wordpress-mobile/WordPress-Android/issues/17905

Therefore I will proceed to merge this PR to the feature branch and make sure the issue you surfaced gets fixed before we merge the whole checkout-related work to `trunk`.

Thank you for reviewing, testing, and helping me with this work @thomashorta  🙇🏻

**Pull request merged**
